### PR TITLE
added ruby

### PR DESCRIPTION
**Pull request by madhura0404**
I need to modify 3rd and 4th exercise to show all the outputs.

### CODE REVIEW

File: Ruby/ex1/1.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+def calculate_serial_average(serial_input)
+    number = serial_input.split('-')
+    xx = number[1].to_f
+    yy = number[2].to_f
+    average = (xx + yy) / 2
+    formatted_average = format('%.2f', average)
+    "#{number[0]}-#{formatted_average}"
+end
+
+class TestSerialAverage < Minitest::Test
```
Comment (minaxi98) [inline Ruby/ex1/1.rb:12]: Check and test cases for invalid input

File: Ruby/ex2/q2.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
+require 'minitest/autorun'
+
+class ArrayManipulator
+    def element_at(arr, index)
+      arr[index]
```
Comment (minaxi98) [inline Ruby/ex2/q2.rb:5]: check for array index out of bound

File: Ruby/ex2/q2.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      arr[start_pos..end_pos]
+    end
+  
+    def non_inclusive_range(arr, start_pos, end_pos)
+      arr[start_pos...end_pos]
+    end
+  
+    def start_and_length(arr, start_pos, length)
+      arr[start_pos, length]
+    end
```
Comment (minaxi98) [inline Ruby/ex2/q2.rb:18]: Array index out of bound check everywhere.

File: Ruby/ex2/q2.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    def test_non_inclusive_range
+      result = @manipulator.non_inclusive_range(@input_array, 1, 3)
+      assert_equal([4, 0], result)
+    end
+  
+    def test_start_and_length
+      result = @manipulator.start_and_length(@input_array, 1, 3)
+      assert_equal([4, 0, -1], result)
+    end
+  end
```
Comment (minaxi98) [inline Ruby/ex2/q2.rb:66]: test cases for edge cases

File: Ruby/ex3/3.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+require 'test/unit'
+
+def hash_manipulator(hash)
+	hash[543121] = "100"
+	hash.each_key do |key|
+		if key.is_a?(Integer)
+			if key.even?
```
Comment (minaxi98) [inline Ruby/ex3/3.rb:7]: can be combined into one if

File: Ruby/ex4/4.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+require 'test/unit'
+
+def skip_sports(sports, skip)
+	return sports[skip..-1] if skip >= 0
```
Comment (minaxi98) [inline Ruby/ex4/4.rb:4]: This is not expected output. Please check the question

File: Ruby/ex6/6.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  def read_and_encode(file_name)
+    content = File.read(file_name)
+    encoded_content = ""
+    bracket = false
+    bracket_content = ""
+  
+    content.each_char do |char|
+      if char == '['
```
Comment (minaxi98) [inline Ruby/ex6/6.rb:22]: Writing reading file can give exception, use exception handling

File: Ruby/ex 5/5.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    loaded_data.delete("name")
+    loaded_data
+  end
+
+  def save_updated_to_json(data)
+    timestamp = Time.now.utc.strftime("%Y-%m-%d_%H-%M-%S")
+    new_file_name = "Madhura_#{timestamp}.json"
+    
+    File.open(new_file_name, "w") do |file|
+      file.write(JSON.pretty_generate(data))
```
Comment (minaxi98) [inline Ruby/ex 5/5.rb:32]: Exception handling in code and test

**Review by minaxi98 — CHANGES_REQUESTED** [review]
